File: Leet_Code_Problems/452.minimum-number-of-arrows-to-burst-balloons.java
/*
 * @lc app=leetcode id=452 lang=java
 *
 * [452] Minimum Number of Arrows to Burst Balloons
 */

// @lc code=start
class Solution {
    public int findMinArrowShots(int[][] points) {

        int len = points.length, i, arrows = 1, prevEnd;
        if (len == 0)
            return 0;
        if (len == 1)
            return 1;

        Arrays.sort(points, (a, b) -> a[1] - b[1]); // Sort with the Ending Diameter

        prevEnd = points[0][1];

        for (i = 0; i < len; i++) {
            if (points[i][0] <= prevEnd)
                continue;
            prevEnd = points[i][1];
            arrows++;
        }

        return arrows;
    }
}
// @lc code=end

File: Leet_Code_Problems/452.minimum-number-of-arrows-to-burst-balloons.cpp
/*
 * @lc app=leetcode id=452 lang=cpp
 *
 * [452] Minimum Number of Arrows to Burst Balloons
 */

// @lc code=start

bool comparator(vector<int> &x, vector<int> &y)
{
    return x[1] < y[1];
}

class Solution
{
public:
    int findMinArrowShots(vector<vector<int>> &points)
    {
        if (points.size() == 0)
            return 0;
        if (points.size() == 1)
            return 1;

        int arrow = 1, i, prevEnd;
        sort(points.begin(), points.end(), comparator);
        prevEnd = points[0][1];

        for (i = 1; i < points.size(); i++)
        {
            // If Curr Start LESS THAN prev end
            if (points[i][0] <= prevEnd) //  Overlapping
                continue;

            arrow++;
            prevEnd = points[i][1];
        }

        return arrow;
    }
};
// @lc code=end
